Allow filters to work in pivot tables

robdaemon opened the pull request:
The `query->expected-cols` function only worked with queries of the type `:query` but `"query"` is perfectly valid also.

Modify the function to take that type, and remove the extraneous calls to `preprocess-query` - the result of a call to
`preprocess-query` cannot be passed to `preprocess-query` a second time without causing a schema violation error.

Commit:
Allow filters to work in pivot tables

The `query->expected-cols` function only worked with queries of the type `:query` but `"query"` is perfectly valid also.

Modify the function to take that type, and remove the extraneous calls to `preprocess-query` - the result of a call to
`preprocess-query` cannot be passed to `preprocess-query` a second time without causing a schema violation error.

paulrosenzweig commented on test/metabase/api/advanced_computation_test.clj:
```diff
@@ -63,6 +71,24 @@
+           (is (= [nil "Organic" 3764] (drop-last (nth rows 91))))
+           (is (= [nil nil 7562] (drop-last (last rows)))))))))
+
 (deftest pivot-card-test
```
Maybe it's worth adding a test of parameters in here.

robdaemon replied:
yep that's my plan! I just didn't get to it yet.

Commit:
Adding parameters test

Commit:
fields and expressions

Make sure that the pivot query works when the source query already includes fields and expressions

Commit:
cleaning up ns declaration

sbelak commented on src/metabase/query_processor.clj:
```diff
@@ -176,7 +176,7 @@
   queries with native source queries won't work, since we don't need the results."
   [{query-type :type, :as query}]
-  (when-not (= query-type :query)
+  (when-not (= (keyword query-type) :query)
```
perhaps `mbql.u/normalize-token` is a more robust choice?

robdaemon replied:
ahhh I didn't even know about that.

sbelak commented on src/metabase/query_processor/pivot.clj:
```diff
@@ -15,16 +15,23 @@
+      (assoc-in [:query :breakout] breakout)
+      ;;TODO: `pivot-grouping` is not "magic" enough to mark it as an internal thing
+      (update-in [:query :fields]
+                 #(conj % [:expression "pivot-grouping"]))
```
I don't think you need a lambda here, update passes the rest of the args along

robdaemon replied:
it's the ordering that's the issue if I don't use a lambda:

`(conj [:foo "bar"] nil)` returns `[:foo "bar" nil]` and I don't want to add a nil if there aren't any `:fields`

sbelak replied:
`mbql.u/update-in-unless-empty` or `m/update-in-when`

sbelak commented on src/metabase/query_processor/pivot.clj:
```diff
@@ -15,16 +15,23 @@
+      (update-in [:query :fields]
+                 #(conj % [:expression "pivot-grouping"]))
+      ;;TODO: replace this value with a bitmask or something to indicate the source better
+      (update-in [:query :expressions]
```
see above

sbelak reviewed (approved):
Looks good. Some minor stylistic nibbles

Commit:
Switch to mbql.u/normalize-token
